### PR TITLE
Introduce gosec for Static Application Security Testing (SAST)

### PR DESCRIPTION
**Pull request by ashwani2k**
**What this PR does / why we need it**:
This PR introduces two make targets: sast and sast-report to run gosec for Static Application Security Testing. Additionally, it also addresses the security vulnerabilities

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement developer
Add `gosec` for Static Application Security Testing (SAST) via `make sast` and `make sast-report`.
```

### CODE REVIEW

File: .ci/check [hunk 1]
```diff
@@ -50,3 +50,6 @@ echo "Running golint..."
 for dir in ${PACKAGES_DIRS}; do
   golint -set_exit_status $(find $dir -maxdepth 1 -name "*.go" | grep -vE 'zz_generated|_test.go')
 done
+
+# Run Static Application Security Testing (SAST) using gosec
+make sast-report
```
Comment (shreyas-s-rao) [inline .ci/check:55]: Please add newline

File: Makefile [hunk 2]
```diff
@@ -134,3 +136,11 @@ CONTROLLER_GEN=$(GOBIN)/controller-gen
...
 CONTROLLER_GEN=$(shell which controller-gen)
 endif
+
+.PHONY: sast
+sast: $(GOSEC)
+	@./hack/sast.sh
+
+.PHONY: sast-report
+sast-report: $(GOSEC)
+	@./hack/sast.sh --gosec-report true
```
Comment (shreyas-s-rao) [inline Makefile:146]: Same as above

File: hack/install-gosec.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  rm -rf "${temp_dir}"
+}
+trap cleanup EXIT ERR INT TERM
+echo "Downloading from: https://github.com/securego/gosec/releases/download/${version}/${file_name}"
+curl -L -o ${temp_dir}/${file_name} "https://github.com/securego/gosec/releases/download/${version}/${file_name}"
+
+
+tar -xzm -C "${temp_dir}" -f "${temp_dir}/${file_name}"
+mv "${temp_dir}/gosec" $TOOLS_BIN_DIR
+chmod +x $TOOLS_BIN_DIR/gosec
```
Comment (shreyas-s-rao) [inline hack/install-gosec.sh:46]: Same as above

File: hack/sast.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  gosec_report_parse_flags="-track-suppressions -fmt=sarif -out=gosec-report.sarif -stdout"
+fi
+
+# KUPID uses code-generators https://github.com/kubernetes/code-generator which create lots of G103 (CWE-242:
+# Use of unsafe calls should be audited) & G104 (CWE-703: Errors unhandled) errors.
+# However, those generators are best-pratice in Kubernetes environment and their results are tested well.
+# Thus, generated code is excluded from gosec scan.
+# Nested go modules are not supported by gosec (see https://github.com/securego/gosec/issues/501), so the ./hack folder
+# is excluded too. It does not contain productive code anyway.
+${TOOLS_BIN_DIR}/gosec -exclude-generated -exclude-dir=hack $gosec_report_parse_flags ./...
```
Comment (shreyas-s-rao) [inline hack/sast.sh:45]: Same as above

File: hack/tools.mk [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+TOOLS_BIN_DIR              := $(TOOLS_DIR)/bin
+
+#Tool Binaries
+GOSEC ?= $(TOOLS_BIN_DIR)/gosec
+
+#Tool Versions
+GOSEC_VERSION ?= v2.21.4
+
+$(GOSEC):
+		GOSEC_VERSION=$(GOSEC_VERSION) bash $(abspath $(TOOLS_DIR))/install-gosec.sh
```
Comment (shreyas-s-rao) [inline hack/tools.mk:14]: Same as above

**Review by shreyas-s-rao — COMMENTED** [review]
@ashwani2k thanks for the PR. Other than the nits, LGTM. PTAL

File: hack/tools.mk (hunk 5 above)
Comment (ashwani2k) [reply hack/tools.mk]: Resovled with [8adf8e6](https://github.com/gardener/kupid/pull/75/commits/8adf8e6ba8ff6fe575243b3edf3e63d73cd19efd)

File: hack/sast.sh (hunk 4 above)
Comment (ashwani2k) [reply hack/sast.sh]: Resolved with [8adf8e6](https://github.com/gardener/kupid/pull/75/commits/8adf8e6ba8ff6fe575243b3edf3e63d73cd19efd)

File: hack/install-gosec.sh (hunk 3 above)
Comment (ashwani2k) [reply hack/install-gosec.sh:46]: Resolved with [8adf8e6](https://github.com/gardener/kupid/pull/75/commits/8adf8e6ba8ff6fe575243b3edf3e63d73cd19efd)

File: Makefile (hunk 2 above)
Comment (ashwani2k) [reply Makefile:146]: Resolved with [8adf8e6](https://github.com/gardener/kupid/pull/75/commits/8adf8e6ba8ff6fe575243b3edf3e63d73cd19efd)

File: .ci/check (hunk 1 above)
Comment (ashwani2k) [reply .ci/check]: Resolved with [8adf8e6](https://github.com/gardener/kupid/pull/75/commits/8adf8e6ba8ff6fe575243b3edf3e63d73cd19efd)

**Review by shreyas-s-rao — APPROVED** [review]
/lgtm

**Pull request merged**